sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -566,6 +568,21 @@ object TypeCoercion {
     }
   }
 
+  /**
+   * When all inputs in [[Concat]] are binary, coerces an output type to binary
+   */
+  case class ConcatCoercion(conf: SQLConf) extends TypeCoercionRule {
```

Comment:
I feel this rule is a little hacky as it must be run before `ImplicitTypeCasts`. I think we should not make `Concat` extends `ImplicitCastInputTypes`, and do type coercion manually for it with a customer rule.

Reply:
Aha, sounds reasonable. I feel we should fix that in an separate pr first, thought? Or, this pr should include the fix, too?

Reply:
I think we should do it in this PR, because this is a new requirement for the new behavior introduced in this PR.

Reply:
ok